pandas/indexes/base.py

```diff
@@ -2366,7 +2365,17 @@ def groupby(self, to_groupby):
         groups : dict
             {group name -> group labels}
         """
-        return self._groupby(self.values, _values_from_object(to_groupby))
+        from pandas.core.groupby import _groupby_indices
+
+        # TODO: if we are a MultiIndex, we can do better
+        # that converting to tuples
+        to_groupby = getattr(to_groupby, 'values', to_groupby)
+        result = _groupby_indices(to_groupby)
+
+        # map to the label
+        result = {k: self.take(v) for k, v in compat.iteritems(result)}
```

Comment:
I'm just wondering whether producing a fully-boxed Index for each group up front has memory use or performance implications. This is something I guess we'll address in much more detail in pandas 2.0